### PR TITLE
Add client statistics support (+ add tests on users & clients statistics)

### PR DESCRIPTION
**Pull request by mfouilleul**

### CODE REVIEW

**Comment by SuperQ** [comment]
Awesome, looks great!

**Comment by SuperQ** [comment]
@juliusv or @beorn7 for some additional golang :eyes:

**Comment by SuperQ** [comment]
Also thanks for cleaning up the SQL query (removing the `SELECT *`) in `info_schema_userstats.go`

**Comment by mfouilleul** [comment]
I have a general golang question for you. In the main mysqld_exporter.go
you import the collector directory with:

import "github.com/prometheus/mysqld_exporter/collector"

As I forked the repo the make is broken because of this path, I should edit
the import path to:
import "github.com/mfouilleul/mysqld_exporter/collector"

Do you have some advises to manage this issue?

Max.

2016-06-20 16:30 GMT+02:00 Ben Kochie <notifications@github.com>:

**Comment by beorn7** [comment]
Yeah, that's a problem. You could work locally in the original repository, but push to your fork, I guess...

File: collector/info_schema_clientstats.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		return err
+	}
+	defer informationSchemaClientStatisticsRows.Close()
+
+	// The client column is assumed to be column[0], while all other data is assumed to be coerceable to float64.
+	// Because of the client column, clientStatData[0] maps to columnNames[1] when reading off the metrics
+	// (because clientStatScanArgs is mapped as [ &client, &clientData[0], &clientData[1] ... &clientdata[n] ]
+	// To map metrics to names therefore we always range over columnNames[1:]
+	var columnNames []string
+	columnNames, err = informationSchemaClientStatisticsRows.Columns()
```
Comment (beorn7) [inline collector/info_schema_clientstats.go]: `columnNames, err :=  ...` should work as well, then you don't need the `var` line above.

File: collector/info_schema_clientstats.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	if err != nil {
+		return err
+	}
+
+	var client string                                        // Holds the client name, which should be in column 0.
+	var clientStatData = make([]float64, len(columnNames)-1) // 1 less because of the client column.
+	var clientStatScanArgs = make([]interface{}, len(columnNames))
```
Comment (beorn7) [inline collector/info_schema_clientstats.go]: Id prefer a `var ( ... )` block with that many `var` lines.

File: collector/info_schema_clientstats.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	clientStatScanArgs[0] = &client
+	for i := range clientStatData {
+		clientStatScanArgs[i+1] = &clientStatData[i]
+	}
+
+	for informationSchemaClientStatisticsRows.Next() {
+		err = informationSchemaClientStatisticsRows.Scan(clientStatScanArgs...)
+		if err != nil {
```
Comment (beorn7) [inline collector/info_schema_clientstats.go]: Write in one row:

`if err := informationSchemaClientStatisticsRows.Scan(clientStatScanArgs...); err != nil {`

**Comment by mfouilleul** [comment]
Ok. Basically I just copied the info_schema_userstats.go and search&replace user/client.

**Comment by SuperQ** [comment]
Don't worry about fixing info_schema_userstats.go in this PR, we can clean it up later.

**Comment by beorn7** [comment]
Yeah, mysqld_exporter is a patchwork of different style violations. We have to clean it up eventually. But we also try to improve new contributions.

File: collector/info_schema_clientstats.go [hunk 4]
```diff
@@ -139,27 +139,32 @@ func ScrapeClientStat(db *sql.DB, ch chan<- prometheus.Metric) error {
 	}
 	defer informationSchemaClientStatisticsRows.Close()
 
-	// The client column is assumed to be column[0], while all other data is assumed to be coerceable to float64.
+	// The client
```
Comment (beorn7) [inline collector/info_schema_clientstats.go]: Did you accidentally truncate this comment line? It seemed meaningful to me.

File: collector/info_schema_clientstats.go [hunk 5]
```diff
@@ -139,27 +139,32 @@ func ScrapeClientStat(db *sql.DB, ch chan<- prometheus.Metric) error {
...
 		clientStatScanArgs[i+1] = &clientStatData[i]
 	}
 
 	for informationSchemaClientStatisticsRows.Next() {
-		err = informationSchemaClientStatisticsRows.Scan(clientStatScanArgs...)
-		if err != nil {
+		// err = informationSchemaClientStatisticsRows.Scan(clientStatScanArgs...)
+		// if err != nil {
+		// 	return err
+		// }
```
Comment (beorn7) [inline collector/info_schema_clientstats.go]: Remove the commented left-over.

**Comment by beorn7** [comment]
OK, I guess that's a :+1: from my side. @SuperQ could you double-check the functional side and merge as you see fit?

**Comment by mfouilleul** [comment]
Still have some issues with this PR?

**Comment by SuperQ** [comment]
Ahh, sorry, this just fell out of my inbox.

**Pull request merged**